File: input/parse.go
package input

import (
	"fmt"
	"net/url"
	"regexp"
)

var (
	numericRe = regexp.MustCompile(`^\d+$`)
	statusRe  = regexp.MustCompile(`/status/(\d+)`)
)

// ExtractTweetID extracts a tweet ID from a raw tweet ID, Twitter/X URL, or Nitter URL.
func ExtractTweetID(input string) (string, error) {
	if numericRe.MatchString(input) {
		return input, nil
	}

	u, err := url.Parse(input)
	if err != nil {
		return "", fmt.Errorf("invalid input %q: %w", input, err)
	}

	m := statusRe.FindStringSubmatch(u.Path)
	if m == nil {
		return "", fmt.Errorf("could not extract tweet ID from %q", input)
	}
	return m[1], nil
}

File: web/src/input/parse.ts
const numericRe = /^\d+$/;
const statusRe = /\/status\/(\d+)/;

/** Extract a tweet ID from a raw tweet ID, Twitter/X URL, or Nitter URL. */
export function extractTweetID(input: string): string {
  input = input.trim();

  if (numericRe.test(input)) {
    return input;
  }

  try {
    const u = new URL(input);
    const m = statusRe.exec(u.pathname);
    if (m) {
      return m[1];
    }
  } catch {
    // not a valid URL
  }

  throw new Error(`could not extract tweet ID from "${input}"`);
}
